### PR TITLE
deploy: allow to trigger deployment when ICT is updated

### PR DESCRIPTION
**Pull request by mfojtik**
Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=1444545

Basically detect if there are any new triggers in the current DC (comparing to previous version). If yes, then don't prevent the controller from triggering.

### CODE REVIEW

**Comment by mfojtik** [comment]
[test]

File: pkg/deploy/registry/instantiate/rest.go [hunk 1]
```diff
@@ -208,12 +208,15 @@ func canTrigger(
 		// change is an image change. Look at the deserialized config's
 		// triggers and compare with the present trigger. Initial deployments
 		// should always trigger - there is no previous config to use for the
-		// comparison.
-		if config.Status.LatestVersion > 0 && !triggeredByDifferentImage(*t.ImageChangeParams, *decoded) {
+		// comparison. Also configs with new/updated triggers should always trigger.
+		if config.Status.LatestVersion == 0 || hasUpdatedTriggers(*config, *decoded) || triggeredByDifferentImage(*t.ImageChangeParams, *decoded) {
```
Comment (mfojtik) [inline pkg/deploy/registry/instantiate/rest.go]: this logic need check, it makes my brain hurt.

**Comment by mfojtik** [comment]
@kargakis I want to add extended test proving this works as expected.

**Comment by kargakis** [comment]
I think we also need to make the generictrigger controller enqueue DCs when there is an update in the triggers of the DC.

File: pkg/deploy/registry/instantiate/rest.go (hunk 1 above)
Comment (kargakis) [inline pkg/deploy/registry/instantiate/rest.go]: looks fine

**Comment by mfojtik** [comment]
@kargakis +1

**Comment by mfojtik** [comment]
[test]

**Comment by mfojtik** [comment]
@kargakis added extended test to prove this works, PTAL

File: test/extended/deployments/deployments.go [hunk 2]
```diff
@@ -307,6 +308,59 @@ var _ = g.Describe("deploymentconfigs", func() {
...
+			_, name, err := createFixture(oc, imageChangeTriggerFixture)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			o.Expect(waitForSyncedConfig(oc, name, deploymentRunTimeout)).NotTo(o.HaveOccurred())
+
+			g.By("tagging the busybox:latest as test:v1 image")
+			_, err = oc.Run("tag").Args("docker.io/busybox:latest", "test:v1").Output()
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			expectLatestVersion := func(version int) {
+				dc, err := oc.Client().DeploymentConfigs(oc.Namespace()).Get(name)
```
Comment (kargakis) [inline test/extended/deployments/deployments.go]: Log this error and return nil

File: test/extended/deployments/deployments.go [hunk 3]
```diff
@@ -307,6 +308,59 @@ var _ = g.Describe("deploymentconfigs", func() {
...
+				latestVersion := dc.Status.LatestVersion
+				err = wait.PollImmediate(500*time.Millisecond, 10*time.Second, func() (bool, error) {
+					dc, err = oc.Client().DeploymentConfigs(oc.Namespace()).Get(name)
+					if err != nil {
+						return false, err
```
Comment (kargakis) [inline test/extended/deployments/deployments.go]: Log this error and return nil

File: test/extended/deployments/deployments.go (hunk 2 above)
Comment (kargakis) [inline test/extended/deployments/deployments.go]: Missing error check

File: test/extended/deployments/deployments.go [hunk 4]
```diff
@@ -307,6 +308,59 @@ var _ = g.Describe("deploymentconfigs", func() {
...
+				}
+				o.Expect(err).NotTo(o.HaveOccurred())
+				o.Expect(waitForLatestCondition(oc, name, deploymentRunTimeout, deploymentReachedCompletion)).NotTo(o.HaveOccurred())
+			}
+
+			g.By("ensuring the deployment config latest version is 1 and rollout completed")
+			expectLatestVersion(1)
+
+			g.By("updating the image change trigger to point to test:v2 image")
+			{
```
Comment (kargakis) [inline test/extended/deployments/deployments.go]: Why these?

**Comment by kargakis** [comment]
> I think we also need to make the generictrigger controller enqueue DCs when there is an update in the triggers of the DC.

Shouldn't we also fix this here?

**Comment by mfojtik** [comment]
> Shouldn't we also fix this here?

will prefer separate PR to unblock customer first.

File: test/extended/deployments/deployments.go [hunk 5]
```diff
@@ -307,6 +308,59 @@ var _ = g.Describe("deploymentconfigs", func() {
...
+				_, err := oc.Run("set").Args("triggers", "dc/"+name, "--remove-all").Output()
+				o.Expect(err).NotTo(o.HaveOccurred())
+				_, err = oc.Run("set").Args("triggers", "dc/"+name, "--from-image", "test:v2", "--auto", "-c", "test").Output()
+				o.Expect(err).NotTo(o.HaveOccurred())
+				o.Expect(waitForSyncedConfig(oc, name, deploymentRunTimeout)).NotTo(o.HaveOccurred())
```
Comment (kargakis) [inline test/extended/deployments/deployments.go]: We should switch `waitForSyncedConfig` to accept a generation so you can pass the generation you get from the last `set triggers` update, for this to be strictly correct.

**Comment by kargakis** [comment]
> will prefer separate PR to unblock customer first.

In the test, you are using `oc tag` for triggering a new rollout. Is this the only expectance here or should we also trigger when the DC is updated (at the point of the DC update, we can start deploying) ?

**Comment by mfojtik** [comment]
@kargakis i would say that you need a ConfigChange trigger to trigger when DC is updated?

**Comment by mfojtik** [comment]
@kargakis fixed, thanks!

File: test/extended/deployments/deployments.go (hunk 2 above)
Comment (mfojtik) [inline test/extended/deployments/deployments.go]: fixed.

**Comment by mfojtik** [comment]
flake: https://github.com/openshift/origin/issues/13650

**Comment by kargakis** [comment]
LGTM for unblocking the customer, let's follow-up with fixing the case where the image already exists

**Comment by mfojtik** [comment]
[merge]

**Comment by mfojtik** [comment]
[merge]

**Comment by mfojtik** [comment]
[merge]

**Comment by mfojtik** [comment]
dnf flake [test]

**Comment by mfojtik** [comment]
[test] dnf flake again

**Comment by brenton** [comment]
flake: #13942

**Comment by brenton** [comment]
[merge]

**Comment by brenton** [comment]
flake: #13943

**Comment by kargakis** [comment]
[merge]

**Comment by kargakis** [comment]
@mfojtik did you create any issue for the follow-up work?

**Pull request merged**